### ISSUE TITLE
Runtime configuration, storage costs

### ISSUE THREAD

**Issue by mhalambek** [near/nearcore#8140]
**Describe the bug**
Neard to overwrites storage costs  related parameters from genesis.json runtime configuration.
After setting runtime storage configuration costs to zero, it's not applied in the actual runtime.
I was able to override that behavior by setting storage costs to free manually in the code.

**Expected behavior**
Neard loads runtime parameters from genesis and sets them to zero.

**Version (please complete the following information):**
nearcore version: 1.29.1
rust version: rustc 1.62.1 (e092d0b6b 2022-07-16)
environment: local

**Additional context**

**Comment by akhi3030** [comment]
Looks like something for the core team.  CC: @mzhangmzz and @Longarithm

**Comment by mhalambek** [comment]
Code which overwrites configuration set in genesis

https://github.com/near/nearcore/blob/813ceb9ff082743367966d828741a63f0237e4e8/core/primitives/src/runtime/parameter_table.rs#L51

cc @mm-near I can make a PR with appropriate fix, but not sure about the approach since this looks very sensitive to me in terms of stability of mainnet and testnet

**Comment by mhalambek** [comment]
There is large discrepancy for this file between 1.29.1 and current master, is there any work done now over that part of nearcore.

**Comment by Longarithm** [comment]
We no longer put runtime config parameters to genesis.json. Parameters in this folder are the source of truth: https://github.com/near/nearcore/tree/ddd9f886a9e00dd39556b4d907c06f9ed3e5ce42/core/primitives/res/runtime_configs

You can add new file `NEW_VERSION.txt` and for all parameters put a line `param_name: old_cost -> 0` there, or remove everything except `parameters.txt` and `parameters_testnet.txt` and set all costs to 0 there. 

Maybe we should raise error/warning if genesis config has `runtime_config` field, because it is not used anymore. Moreover, we should remove this field from `utils/mainnet-res/res/mainnet_genesis.json`. @jakmeier what do you think?

**Comment by jakmeier** [comment]
Hm, my understanding is that if genesis_config.json contains any parameters, these are for the genesis block of a particular chain. Nearcore master assumes that there have been several protocol upgrades after that, so it seems consistent to me that the numbers are different from the newest protocol version.

> Maybe we should raise error/warning if genesis config has runtime_config field, because it is not used anymore. Moreover, we should remove this field from utils/mainnet-res/res/mainnet_genesis.json. @jakmeier what do you think?

I don't know what our current practice is here. If we want to keep the mainnet_genesis.json up to date with the latest protocol version, then yeah we should just remove it. But if we want to freeze it at `"genesis_time": "2020-07-21T16:55:51.591948000Z"` and `"protocol_version": 29`, which I thought was the case, then it seems to me we should keep it as is and not produce a warning.

**Comment by chefsale** [comment]
cc: @mm-near genesis overrides are quite important for us at calimero, as we need to zero out storage cost and some other things. @mhalambek is happy to implement the change as we already have it in a fork.

**Comment by mhalambek** [comment]
For Calimero private shards to works as expected without us maintaing a fork we would need to set storage costs as a runtime parameter, would you prefer a explict cmd flag or setting those via genesis.json as it was the case before.

Any other suggestions to set it as runtime parameter are welcome.

**Comment by mhalambek** [comment]
How about providig explicit flag to the run command
which would load free storage runtime params

```bash
neard run --calimero-zero-storage
```

That should prevent any accidents in terms of runnong testnet or mainnet nodes

**Comment by jakmeier** [comment]
Feels hacky to me to add flags for specific projects and then spread if conditions around the code to respect it. And even if you specify it this way, it's not clear to me how you want to pass it from neard CLI down to all places that care about storage costs.

**Comment by mhalambek** [comment]
When it comes to enabling zero costs I went directly to the parameters.txt
and hardcoded it to 0 for sake of testing further down the line

It seems to me that we would need to set storage_costs to 0 consistently across all protocol version for sake of future compatibility. In code that would be here.

https://github.com/calimero-is-near/nearcore/commit/75d28392c6aa736ae4601463b62ff6944e80c146

If chain_id is forwarded to the RuntimeConfigStore I can set storage_costs to zero for all versions available there.
Calimero chain_ids always end in calimero or calimero-testnet

**Comment by mzhangmzz** [comment]
> We no longer put runtime config parameters to genesis.json. Parameters in this folder are the source of truth: https://github.com/near/nearcore/tree/ddd9f886a9e00dd39556b4d907c06f9ed3e5ce42/core/primitives/res/runtime_configs
> 
> You can add new file `NEW_VERSION.txt` and for all parameters put a line `param_name: old_cost -> 0` there, or remove everything except `parameters.txt` and `parameters_testnet.txt` and set all costs to 0 there.
> 
> Maybe we should raise error/warning if genesis config has `runtime_config` field, because it is not used anymore. Moreover, we should remove this field from `utils/mainnet-res/res/mainnet_genesis.json`. @jakmeier what do you think?

Unfortunately we can't remove unused fields from genesis file but we can't change the genesis file for testnet and mainnet, which are already running.

**Comment by chefsale** [comment]
So, what is your proposal on how should we go with implementing this? That information would be helpful for us, as we need this feature?

**Comment by mzhangmzz** [comment]
@chefsale would what @Longarithm suggested up there work for you? Remove all the version files and set the costs to 0 in `parameters.txt` and `parameters_testnet.txt`?

**Comment by chefsale** [comment]
What happens in that case if there is a new release, wont that override the storage cost?

**Comment by jakmeier** [comment]
> It seems to me that we would need to set storage_costs to 0 consistently across all protocol version for sake of future compatibility. In code that would be here.
> 
> https://github.com/near/nearcore/blob/6ae192b44dff2fd3ef922bd5086830082f5d59d2/core/primitives/src/runtime/config_store.rs#L52
> If chain_id is forwarded to the RuntimeConfigStore I can set storage_costs to zero for all versions available there.
Calimero chain_ids always end in calimero or calimero-testnet

For lack of a better alternative, I suggest you create a PR that does this. Long-term, I believe we want better chain-specific configuration. But it's unclear to me how exactly it should look, so maybe starting out with a hack that looks at the chain id and sets some hard-coded zeroes is really the best way to get it started. (Please ping me on a potential PR so I can create an issue for cleaning it up down the road.)

Does that work for you to get you unblocked?

**Comment by mhalambek** [comment]
Have a PR for this, which looks like the best solution so far
Instead of verifying chain_id we use rust feature flags to enable zero storage

https://github.com/near/nearcore/pull/8212

There are plenty of failing tests, but looks pretty solid to me
Tests should be more or less simple to fix

**Comment by jakmeier** [comment]
Yeah I guess cargo flags also work here. If you can fix the tests and maybe add some comments, I would tentatively say we can review and merge your solution to near/nearcore.

In your comments, please also mention https://github.com/near/nearcore/issues/8217
